torch/_dynamo/functional_export.py

```diff
@@ -618,6 +619,235 @@ def normalize_graph_module(gm):
+
+            args, kwargs = self.input_processor(inputs)
+            try:
+                out.forward_callable(
+                    compiled_fn=backend_dummy, extra_globals=f_globals
+                )(*args, **kwargs)
+            except Yield:
+                assert self.gm_inputs is not None
+                return self.gm_inputs
+            raise RuntimeError
+
+    class DynamoBytecodeUnflatten:
+        def __init__(self, input_processor):
+            self.input_processor = input_processor
+
+        def __call__(self, flat_outs, inputs):
+            def backend_dummy(*example_inputs):
+                return flat_outs
+
+            args, kwargs = self.input_processor(inputs)
+            with torch._C._DisableTorchDispatch():
+                results = out.forward_callable(
+                    compiled_fn=backend_dummy, extra_globals=f_globals
+                )(*args, **kwargs)
+            return results
+
+    graph_module = backend_input.graph_module
+    if isinstance(root, torch.nn.Module):
+        graph_module._parameters = root._parameters.copy()
+        graph_module._buffers = root._buffers.copy()
+        assert all(not hasattr(graph_module, m) for m in root._modules)
+        graph_module._modules.update(root._modules)
+        graph_module._non_persistent_buffers_set = (
+            root._non_persistent_buffers_set.copy()
+        )
+        if sys.version_info >= (3, 14):
+            import annotationlib  # added in 3.14
+
+            annotations = annotationlib.get_annotations(torch.nn.Module)
+        else:
+            annotations = getattr(torch.nn.Module, "__annotations__", None)
+        for name, value in root.__dict__.items():
+            if annotations and name not in annotations:
+                graph_module.__dict__[name] = value
+        graph_module._forward_hooks = root._forward_hooks.copy()
+        graph_module._forward_pre_hooks = root._forward_pre_hooks.copy()
+        graph_module._backward_hooks = root._backward_hooks.copy()
+        graph_module._backward_pre_hooks = root._backward_pre_hooks.copy()
+        if graph_module._forward_hooks or graph_module._forward_pre_hooks:
+            assert getattr(graph_module, "_wrapped_call", None) is not None
+            assert isinstance(
+                graph_module._wrapped_call, torch.fx.graph_module._WrappedCall
+            )
+            assert graph_module._wrapped_call.cls_call is None
+
+            def dynamo_wrapped_call(self, *args, **kwargs):
+                assert "forward" not in self.__dict__
+
+                fwd_hooks = self._forward_hooks
+                fwd_pre_hooks = self._forward_pre_hooks
+                original_forward = type(self).forward
+
+                def patched_forward(self, *args, **kwargs):
+                    self._forward_hooks = fwd_hooks
+                    self._forward_pre_hooks = fwd_pre_hooks
+                    return original_forward(self, *args, **kwargs)
+
+                try:
```

Comment:
Can you pls add inline comments what is happening here?

Reply:
Are you just trying to preserve the forward hooks because for some reason dynamo doesn't trace them?